flink-yarn/src/test/java/org/apache/flink/yarn/TestingYarnAMRMClientAsync.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.yarn;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.api.java.tuple.Tuple4;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.yarn.api.records.ContainerId;
+import org.apache.hadoop.yarn.api.records.Priority;
+import org.apache.hadoop.yarn.api.records.Resource;
+import org.apache.hadoop.yarn.client.api.AMRMClient;
+import org.apache.hadoop.yarn.client.api.async.AMRMClientAsync;
+import org.apache.hadoop.yarn.client.api.async.impl.AMRMClientAsyncImpl;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * A Yarn {@link AMRMClientAsync} implementation for testing.
+ */
+public class TestingYarnAMRMClientAsync extends AMRMClientAsyncImpl<AMRMClient.ContainerRequest> {
```

Comment:
`AMRMClientAsyncImpl` is annotated as unstable. I'm not sure how will the testing implementation works across different Yarn versions. Have we tried this out?

Reply:
Same here.

I'm aware `AMRMClientAsyncImpl` is annotated as `Unstable`. But we are only overriding the minimum set of public methods here. If these methods are changed, then neither the original testing codes nor the production codes of Flink could work anymore.

I have not tried all the Yarn versions. But this should at least work with hadoop 2.4.1 (the default hadoop version if build from sources locally) and 2.8.3 (hadoop version for travis ci tests). I can further verify it with other versions.

Reply:
I think my problem with this class is that we are partially overriding an implementation class here. We don't really know how the overridden methods are used internally. Looking at the code, it looks as if they are all forwarding the calls to `client`. From this perspective it might look fine but if you take a look at `getAvailableResources`, which has not been overridden, then it also calls `client`. Do we know that there is no contract between the overridden methods and `getAvailableResources`? What if the `client` returns different available resources depending on how many container requests have been added? I'm not saying that this is the case, but I want to make that overriding individual methods of an implementation class can cause failures which are hard to predict and even harder to debug.

Reply:
I think we cannot guarantee that there's no contract between the overridden methods and the not-overridden methods. Even we can, these internal things might change in future.

On the other hand, what we can guarantee is that all the methods used by Flink are overridden. It won't be a problem if an un-overridden method is internally related to the overridden methods but is never called.

To that end, I guess we need to make the following changes.
- Override all the methods used in Flink. That means `setHeartbeatInterval` and `unregisterApplicationMaster` in addition to those already overridden.
- Explicitly state in the `javadoc` that all methods used in Flink should be overridden. That might help debug the tests if later Flink uses more public API of `AMRMClientAsync`.

Reply:
I think this is the price we have to pay for integrating with another system. It is always painful and tedious to set up the tests. In parts, this always involves reimplementing the other systems behaviour in mocks. I don't see a better atm. Hence, +1 for what you've proposed.